src/main/java/com/aws/greengrass/authorization/WildcardVariableTrie.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.authorization;
+
+import com.aws.greengrass.util.DefaultConcurrentHashMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Wildcard trie node which contains properties to identify the Node and a map of all it's children.
+ * - isTerminal: If the node is a terminal node while adding a resource. It might not necessarily be a leaf node as we
+ *   are adding multiple resources having same prefix but terminating on different points.
+ * - isTerminalLevel: If the node is the last level before a valid use "#" wildcard (eg: "abc/123/#", 123/ would be the
+ *   terminalLevel).
+ * - isWildcard: If current Node is a valid glob wildcard (*)
+ * - isWildcard: If current Node is a valid MQTT wildcard (#, +)
+ * - matchAll: if current node should match everything. Could be MQTTWildcard or a wildcard and will always be a
+ *   terminal Node.
+ */
+public class WildcardVariableTrie {
+    protected static final char GLOB_WILDCARD = '*';
+    protected static final char MQTT_MULTILEVEL_WILDCARD = '#';
+    protected static final char MQTT_SINGLELEVEL_WILDCARD = '+';
```

Comment:
Create const strings for all of these, then use ex. `GLOB_WILDCARD_STR.charAt(0)` to get the character. This saves us memory and lets you use the proper type without needing type conversions later on (no more String.valueOf())

Reply:
Makes sense, I'll change that!